docs/proposals/provider-agnostic-dns-health-checks.md

```diff
@@ -0,0 +1,116 @@
+# Provider agnostic DNS Health checks
+
+## Introduction
+
+The MGC has the ability of extending the DNS configuration of the gateway with the DNSPolicy resource. This resource allows users to configure health checks. As a result of configuring health checks, the controller creates the health checks in Route53, attaching them to the related DNS records. This has the benefit of automatically disabling an endpoint if it becomes unhealthy, and enabling it again when it becomes healthy again.
+This feature has a few shortfalls:
+
+1. It’s tightly coupled with Route53. If other DNS providers are supported they must either provide a similar feature, or health checks will not be supported
+2. Lacks the ability to reach endpoints in private networks
+
+This document describes a proposal to extend the current health check implementation to overcome these shortfalls.
+
+### Goals
+
+* Ability to configure health checks in the DNSPolicy associated to a Gateway
+* Ability for the health checks to reach endpoints in private networks
+* DNS records are disabled when the associated health check fails
+* Current status of the defined health checks is visible to the end user
+* Transparently keep support for other health check providers like Route53
+
+## Proposal
+
+### Control plane
+
+#### `DNSHealthCheck` resource
+
+The DNSHealthCheck resource represents the configuration of a health check. The reconciliation of this resource results in the creation of health checks in the selected DNS provider, or, if configured to use the agent, the creation of a health check probe in the downstream cluster
+
+```yaml
+apiVersion: kuadrant.io/v1alpha1
+kind: DNSHealthCheck
+metadata:
+  name: example-dnshealthcheck
+  ownerReference:
+    type: DNSPolicy
+    name: example-policy
+spec:  
+  config:
+    port: "..."
+    path: "..."
+    protocol: "..."
+    host: "..."
+    ipAddress: "..."
+  provider: "aws | agent | ..."
+  providerSpecific:
+    "agent/managed-cluster": "..."
+status:
+  providerSpecific:
+    "agent/health-status": "Healthy"
+    "agent/last-check": "..."
+    "aws/health-check-id": "..."
+```
```

Comment:
Mentioned this on the original doc, but do we want to keep the spec of this consistent with DNSRecord to which it is tied quite tightly?  

My understanding of what you are proposing is that we create a DNSHealthCheck resource for every endpoint in a DNSRecord, but we could also create a DNSHealthCheck that contains a list of endpoints the same as the DNSRecord.